argot_test.go

```diff
@@ -17,3 +21,119 @@ func TestStepsSatisftyStep(t *testing.T) {
 		t.Errorf("Unexpected error result from Steps.Go(): %s", err)
 	}
 }
+
+func TestThatNilWorks(t *testing.T) {
+	Steps([]Step{
+		ExpectNil(nil),
+		ExpectError(ExpectNil(42)),
+	}).Test(t)
+}
+
+func TestThatDeepEqualWorks(t *testing.T) {
+	Steps([]Step{
+		ExpectDeepEqual("a", "a"),
+		ExpectError(ExpectDeepEqual("a", "b")),
+	}).Test(t)
+}
+
+func TestThatDiffWorks(t *testing.T) {
+	Steps([]Step{
+		ExpectDiffEqual("a", "a"),
+		ExpectError(ExpectDeepEqual("a", "b")),
+		ExpectDiffEqual(ExpectDeepEqual("a", "b").Go().Error(), `Expected "a" (string), got "b" (string)`),
+	}).Test(t)
+}
+
+func TestThatPrettyWorks(t *testing.T) {
+	Steps([]Step{
+		ExpectPrettyEqual("a", "a"),
+		ExpectError(ExpectPrettyEqual("a", "b")),
+		ExpectPrettyEqual(ExpectPrettyEqual("a", "b").Go().Error(), `Expected equal values, got diff: (-have +want)
+-"a"
++"b"`),
+	}).Test(t)
+}
+
+func TestThatDeferredWorks(t *testing.T) {
+	var a string
+
+	Steps([]Step{
+		StepFunc(func() error {
+			a = "foo"
+			return nil
+		}),
+		NewStepProducer(func() Step {
+			return ExpectDiffEqual(a, "foo")
```

Comment:
I mean, sure, this tests that it works, but it doesn't motivate it - i.e. you could just write
```go
StepFunc(func() error {
  return ExpectDiffEqual(a, "foo")
})
```
the capture of `a` is still delayed here... so I'm still struggling to understand why you want this feature...

Reply:
Your example doesn't work because `ExpectDiffEqual` returns a `Step`, not an `error`. The `StepProducer` adapts that and secondly ensures that the step is only created once.

I've pushed an update that removes some of the extra work, might just be that people expect that the function is only called once, while in fact it might be multiple times in case `String()` is called for the formatting later.